src/main/java/com/percero/agents/sync/dao/DAORegistry.java

```diff
@@ -0,0 +1,39 @@
+package com.percero.agents.sync.dao;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import com.percero.agents.sync.services.DAODataProvider;
+import com.percero.agents.sync.services.DataProviderManager;
+import com.percero.framework.vo.IPerceroObject;
+
+import edu.emory.mathcs.backport.java.util.Collections;
+
+public class DAORegistry {
+	
+	private static DAORegistry instance = null;
+	
+	public static DAORegistry getInstance() {
+		if (instance == null) {
+			instance = new DAORegistry();
+		}
+		return instance;
+	}
+
+	public DAORegistry() {
+		instance = this;
+		
+		DataProviderManager.getInstance().addDataProvider(DAODataProvider.getInstance()); 
+	}
+
+	@SuppressWarnings({ "unchecked" })
+	private static Map<String, IDataAccessObject<? extends IPerceroObject>> dataAccessObjects = Collections.synchronizedMap(new HashMap<String, IDataAccessObject<? extends IPerceroObject>>());
+	
+	public void registerDataAccessObject(String name, IDataAccessObject<? extends IPerceroObject> dataAccessObject) {
+		dataAccessObjects.put(name, dataAccessObject);
+	}
+	
+	public IDataAccessObject<? extends IPerceroObject> getDataAccessObject(String name) {
```

Comment:
Again, you're right. I did some research this time and indeed it looks like you cannot get the generic type at runtime. But you can just pass the Class type in when you register.  It still saves you the cast each time you access a DAO.
```java
public <T extends IPerceroObject> void registerDataAccessObject(Class<T> typeClass, IDataAccessObject<T> dao){
  dataAccessObjects.put(typeClass, dao);
}
```